### ISSUE TITLE
[mono][wasi] Optional wasm-opt pass

### ISSUE THREAD

**Issue by lewing** [dotnet/runtime#94804]

**Comment by RReverser** [comment]
Note that you (.NET) are currently using Clang for linking, which already implicitly runs `wasm-opt` if 1) it's detected on the PATH and 2) linker optimisation is enabled.

It can be enabled via e.g. this in `.csproj`:

```xml
<_WasiSdkClangArgs Include="-O2" />
```

The problem is that currently somewhere in the pipeline .NET either doesn't emit correct [`target_features` custom section](https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md#target-features-section), or maybe strips it. That secion is important for any WebAssembly tools that want to modify the binary as otherwise they don't know which features they can legally use without breaking compatibility. Correspondingly, when enabling link-time optimisation like above, wasm-opt fails validation with errors like:

```
  Performing WASI SDK build: "C:\Users\me\.wasi-sdk\wasi-sdk-20\bin\clang.exe" "@..."
  [wasm-validator error in function 5099] unexpected false: Bulk memory operations require bulk memory [--enable-bulk-memory], on 
  (memory.copy
   (local.get $0)
   (local.get $1)
   (local.get $2)
  )
  [wasm-validator error in function 5100] unexpected false: Bulk memory operations require bulk memory [--enable-bulk-memory], on
  (memory.copy
   (local.get $0)
   (local.get $1)
   (local.get $2)
  )
  [wasm-validator error in function 5101] unexpected false: Bulk memory operations require bulk memory [--enable-bulk-memory], on
  (memory.fill
   (local.get $0)
   (local.get $1)
   (local.get $2)
  )
  Fatal: error validating input
clang : error : linker command failed with exit code 1 (use -v to see invocation)
```

That is, Wasm uses bulk memory operations but doesn't have a `target_features` section that allows them.

**Comment by lewing** [comment]
I'm not sure why the section isn't being generated will take a look. At the moment --enable-bulk-memory and  --enable-simd (If you pass WasmEnableSIMD) are the only post MVP features in the wasi build.

**Comment by RReverser** [comment]
> At the moment --enable-bulk-memory and --enable-simd (If you pass WasmEnableSIMD) are the only post MVP features in the wasi build.

Yeah they're currently passed explicitly, but the thing is that explicitly enabling features from command line really shouldn't be necessary if `target_features` is preserved correctly.

**Comment by pavelsavara** [comment]
Note that because we now produce WASI components, wasm-opt can't be used on those until https://github.com/WebAssembly/binaryen/issues/6728
